rskj-core/src/main/java/org/ethereum/db/IndexedBlockStore.java

```diff
@@ -228,11 +254,13 @@ public synchronized void saveBlock(Block block, BlockDifficulty cummDifficulty,
         blockInfo.setHash(block.getHash().getBytes());
         blockInfo.setMainChain(mainChain);
 
-        if (blocks.get(block.getHash().getBytes()) == null) {
-            blocks.put(block.getHash().getBytes(), block.getEncoded());
+        Optional<byte[]> existingValue = getDSValue(block.getHash().getBytes());
```

Comment:
You should take advantage of chaining and handle errors at the end:

```java
Optional<Block> existingBlock = getDSValue(block.getHash().getBytes()).flatMap(blockStoreEncoder::decodeBlock);
if (!existingBlock.isPresent()) {
    // ...
```

Reply:
nice catch!